flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -287,6 +291,23 @@ public void onUnknownDeploymentsOf(
 
         log.info("Initializing job {} ({}).", jobName, jid);
 
+        if (jobMasterConfiguration.getConfiguration().get(JobManagerOptions.SCHEDULER_MODE)
+                == SchedulerExecutionMode.REACTIVE) {
+            Preconditions.checkState(
+                    slotPoolServiceSchedulerFactory.getSchedulerType()
+                            == JobManagerOptions.SchedulerType.Adaptive,
+                    "Expecting Adaptive Scheduler for reactive mode");
+            log.info("Modifying job parallelism for running in reactive mode.");
+            for (JobVertex vertex : jobGraph.getVertices()) {
+                if (vertex.getMaxParallelism() == JobVertex.MAX_PARALLELISM_DEFAULT) {
+                    vertex.setParallelism(Transformation.UPPER_BOUND_MAX_PARALLELISM);
+                    vertex.setMaxParallelism(Transformation.UPPER_BOUND_MAX_PARALLELISM);
+                } else {
+                    vertex.setParallelism(vertex.getMaxParallelism());
+                }
+            }
```

Comment:
I am wondering whether this logic should rather go into the `DefaultJobManagerRunnerFactory`. That's also where we pick the scheduler implementation. Maybe grouping this logic together makes sense (selecting the right scheduler implementation for the required execution mode and configuring the job graph accordingly). It is more of an idea to discuss. We could also tackle this as a follow up.